jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/ComplianceViolation.java

```diff
@@ -75,13 +81,188 @@ interface Mode
         Set<? extends ComplianceViolation> getAllowed();
     }
 
+    record Event(ComplianceViolation.Mode mode, ComplianceViolation violation, String details)
+    {
+        @Override
+        public String toString()
+        {
+            return String.format("%s (see %s) in mode %s for %s",
+                violation.getDescription(), violation.getURL(), mode, details);
+        }
+    }
+
     /**
      * A listener that can be notified of violations.
      */
     interface Listener
     {
+        Listener NOOP = new Listener() {};
+
+        /**
+         * Initialize the listener in preparation for a new request life cycle.
+         * @return The {@link Listener} instance to use for the request life cycle.
+         */
+        default Listener initialize()
+        {
+            return this;
+        }
+
+        /**
+         * A new Request has begun.
+         *
+         * @param request the request attributes, or null if the Request does not exist yet (eg: during parsing of HTTP/1.1 headers, before request is created)
+         */
+        default void onRequestBegin(Attributes request)
+        {
+        }
+
+        /**
+         * A Request has ended.
+         *
+         * @param request the request attribtues, or null if Request does not exist yet (eg: during handling of a {@link BadMessageException})
+         */
+        default void onRequestEnd(Attributes request)
+        {
+        }
+
+        /**
+         * The compliance violation event.
+         *
+         * @param event the compliance violation event
+         */
+        default void onComplianceViolation(Event event)
+        {
+            onComplianceViolation(event.mode, event.violation, event.details);
+        }
+
+        /**
+         * The compliance violation event.
+         *
+         * @param mode the mode
+         * @param violation the violation
+         * @param details the details
+         * @deprecated use {@link #onComplianceViolation(Event)} instead.  Will be removed in Jetty 12.1.0
+         */
+        @Deprecated(since = "12.0.5", forRemoval = true)
         default void onComplianceViolation(Mode mode, ComplianceViolation violation, String details)
         {
         }
     }
+
+    /**
+     * A Listener that represents multiple user {@link ComplianceViolation.Listener} instances
+     */
+    class ListenerCollection implements Listener
```

Comment:
We really don't want all these new classes to be `public`, as they are just an implementation detail.